README.md
# jQuery Timepicker [![Build Status](http://img.shields.io/travis/wvega/timepicker.svg)](https://travis-ci.org/wvega/timepicker) [![Gitter chat](https://badges.gitter.im/wvega/timepicker.png)](https://gitter.im/wvega/timepicker)

A jQuery plugin to enhance standard form input fields helping users to select
(or type) times.

Please visit http://timepicker.co for Documentation, Examples and Getting Started information.

## Release History

See [CHANGELOG](https://github.com/wvega/timepicker/blob/master/CHANGELOG).

## License
Copyright (c) 2012 Willington Vega
Licensed under the MIT, GPL licenses.


CHANGELOG
* Sat Jul 9 2016 - Version 1.3.5

- Fix stray comma in bower.json (#81).

* Sun Jun 26 2016 - Version 1.3.4

- Allow jquery-timepicker to work with Browserify (#75).
- Add Bower support thanks to @madalinignisca (package name: jquery-timepicker-wvega) (#74).

* Fri May 20 2016 - Version 1.3.3
- Prevent dropdown from closing in IE when user clicks scrollbar (#56).
- Prevent long-click from closing the timepicker (#50).

* Sat Sep 13 2014 - Version 1.3.2
- Add defaultTime option (#41).
- Fix position bug (#43).

* Sat Nov 09 2013 - Version 1.3.1
- Fix minor errors in jQuery Package Manifest.

* Sat Nov 09 2013 - Version 1.3.0
- Drop support for jQuery 1.4.2 or below. Plugin requires jQuery 1.6.0 or newer.
- Fix: Avoid showing dupliate time entries (#21)
- Fix: next, previous, open and close methods are now chainable (#23)
- Fix: getTime now takes into account values set using $.fn.val (#24)
- Fix: Parse values like 12:77 into a valid datetime entry (#27).
- Fix: Parse values like 6666 into a valid datetime entry (#33).
- Fix: Use default cursor in time selector dropdown (#34).
- Fix: Remove ReferenceError: event is not defined (#39).
- Added lowercase am/pm format.
- time-change event and change callbacks are triggered even if the new
  time is null.
- Minor CSS fixes.

* Sun Feb 26 2012 - Version 1.2.2
- Fix too much recursion error in parseTime function (#20).
- Verifies argument passed to parseTime is string.
- Add support for jQuery 1.6 prop function.

* Mon Nov 14 2011 - Version 1.2.1
- Fix strings support for startTime, minTime and maxTime options.

* Sun Nov 13 2011 - Version 1.2.0
- Add support for calling API methods using the plugin function,
  passing the method's name as first argument and options as
  subsequent arguments.
- Add an option() method to allow changing options values after
  initialization.
- Add dropdown option to control whether the dropdown is displayed or not.
- Add scrollbar option to control whether the scrollbars are displayed or not.
- Aceept strings as values for the minTime, maxTime and startTime options.

* Tue Jul 26 2011 - Version 1.1.2
- Released under MIT and GPL Version 2 licenses.
- Add zindex option to override input field's z-index (thanks to Ian Enders).

* Tue Nov 23 2010 - Version 1.1.1
- Fix Issue #11 (https://github.com/wvega/timepicker/issues/11)

* Mon Nov 08 2010 - Version 1.1.0
- Use input field's z-index.
- Do not use rounded corners by default.
- Use a single UL element for all timepicker in the same page.
- Interval greather than 60 minutes are now supported.
- Add time-change custom event.

* Fri Oct 29 2010 - Version 1.0.3
- Add dynamic option

* Fri Oct 08 2010 - Version 1.0.2
- Add support for jQuery 1.3 and superior (thanks to Brian Link)
- Fix bug with jQuery 1.4.3

* Thu Oct 07 2010 - Version 1.0.1
- Fix selection issue in some versions of IE.
  http://github.com/wvega/timepicker/issues/closed/#issue/3

* Wed Jul 17 2010 - Version 1.0.0
- jQuery Timepicker official release


test/test.js
/*global module, test, expect */
/*global ok, equal, deepEqual, start, stop */

/**
 * ======== A Handy Little QUnit Reference ========
 * http://docs.jquery.com/QUnit
 * Test methods:
 *   expect(numAssertions)
 *   stop(increment)
 *   start(decrement)
 * Test assertions:
 *   ok(value, [message])
 *   equal(actual, expected, [message])
 *   notEqual(actual, expected, [message])
 *   deepEqual(actual, expected, [message])
 *   notDeepEqual(actual, expected, [message])
 *   strictEqual(actual, expected, [message])
 *   notStrictEqual(actual, expected, [message])
 *   raises(block, [expected], [message])
 */

/* jQuery 1.4 queue - needed when testing with previos versions of jQuery */

// XXX: I think the tests can be rewritten to avoid using jQuery's queue
if (jQuery.fn.jquery < '1.4') {
    (function(jQuery) {

        jQuery.extend({
            queue: function( elem, type, data ) {
                if ( !elem ) {
                    return;
                }

                type = (type || 'fx') + 'queue';
                var q = jQuery.data( elem, type );

                // Speed up dequeue by getting out quickly if this is just a lookup
                if ( !data ) {
                    return q || [];
                }

                if ( !q || jQuery.isArray(data) ) {
                    q = jQuery.data( elem, type, jQuery.makeArray(data) );

                } else {
                    q.push( data );
                }

                return q;
            },

            dequeue: function( elem, type ) {
                type = type || 'fx';

                var queue = jQuery.queue( elem, type ), fn = queue.shift();

                // If the fx queue is dequeued, always remove the progress sentinel
                if ( fn === 'inprogress' ) {
                    fn = queue.shift();
                }

                if ( fn ) {
                    // Add a progress sentinel to prevent the fx queue from being
                    // automatically dequeued
                    if ( type === 'fx' ) {
                        queue.unshift('inprogress');
                    }

                    fn.call(elem, function() {
                        jQuery.dequeue(elem, type);
                    });
                }
            }
        });

        jQuery.fn.extend({
            queue: function( type, data ) {
                if ( typeof type !== 'string' ) {
                    data = type;
                    type = 'fx';
                }

                if ( data === undefined ) {
                    return jQuery.queue( this[0], type );
                }
                return this.each(function(/*i*/) {
                    var queue = jQuery.queue( this, type, data );

                    if ( type === 'fx' && queue[0] !== 'inprogress' ) {
                        jQuery.dequeue( this, type );
                    }
                });
            },
            dequeue: function( type ) {
                return this.each(function() {
                    jQuery.dequeue( this, type );
                });
            },

            // Based off of the plugin by Clint Helfers, with permission.
            // http://blindsignals.com/index.php/2009/07/jquery-delay/
            delay: function( time, type ) {
                time = jQuery.fx ? jQuery.fx.speeds[time] || time : time;
                type = type || 'fx';

                return this.queue( type, function() {
                    var elem = this;
                    setTimeout(function() {
                        jQuery.dequeue( elem, type );
                    }, time );
                });
            },

            clearQueue: function( type ) {
                return this.queue( type || 'fx', [] );
            }
        });

    })(jQuery);
}


/* jQuery Timepicker Tests */
(function($) {
    $(function() {
        var system = $({}),
            delay = 100, // ms
            teardown;

        teardown = function() {
            var instance = $('#timepicker').data('TimePicker');

            if (instance) {
                instance.destroy();
            }
        };


        module('TimePicker API', { teardown: teardown });

        test('timepicker instance', function() {
            var timepicker = $('#timepicker').timepicker(),
                instance = timepicker.timepicker();
            ok(typeof instance.widget !== 'undefined', 'Get access to the instance object.');
        });

        test('timepicker destroy', function() {
            var timepicker = $('#timepicker').timepicker();
            ok(timepicker.data('TimePicker') !== null, 'TimePicker instance exists.');
            timepicker.timepicker().destroy();
            ok(timepicker.data('TimePicker') === null, 'TimePicker instance no longer exists.');
        });

        test('chainable methods: next, previous open, close, destroy', function() {
            var timepicker = $('#timepicker').timepicker(),
                instance = timepicker.timepicker(),
                object = null;

            object = instance.open();
            ok(object.jquery, 'open() method returns a jQuery instance.');

            object = timepicker.timepicker('next');
            ok(object.jquery, 'next() method returns a jQuery instance.');

            object = timepicker.timepicker('previous');
            ok(object.jquery, 'previous() method returns a jQuery instance.');

            object = instance.close();
            ok(object.jquery, 'close() method returns a jQuery instance.');

            object = instance.destroy();
            ok(object.jquery, 'destroy() method returns a jQuery instance.');
        });

        test('selected, first, last', function() {
            var timepicker = $('#timepicker').timepicker(),
                instance = timepicker.timepicker(),
                selected = false,
                expected = 5;

            expect(expected);
            stop();

            system.queue('test', []);
            system.queue('test', function(next) {
                selected = instance.selected();
                ok(selected === null, 'No item is selected at the beginning!.');
                timepicker.simulate('keydown', {keyCode: $.TimePicker.prototype.keyCode.DOWN});
                next();
            });

            system.delay(delay, 'test');
            system.queue('test', function(next) {
                selected = instance.selected();
                ok(selected !== null, 'An element is selected after the DOWN key is pressed.');
                ok(instance.first(), 'That element is the first element.');
                timepicker.simulate('keydown', {keyCode: $.TimePicker.prototype.keyCode.UP});
                next();
            });

            system.delay(delay, 'test');
            system.queue('test', function(next) {
                selected = instance.selected();
                ok(selected !== null, 'Another element is selected after the UP key is pressed.');
                ok(instance.last(), 'That element is the last element.');
                next();
            });

            system.queue('test', function(/*next*/) { start(); }).dequeue('test');
        });

        test('parse', function() {
            var timepicker = $('#timepicker').timepicker(),
                instance = timepicker.timepicker(),
                now = new Date(),
                value, parsed, result, expected, message, k, n, t;

            function time(hours, minutes, seconds) {
                hours = hours || 0;
                minutes = minutes || 0;
                seconds = seconds || 0;

                var t = new Date();
                t.setTime(now.valueOf());
                t.setHours(hours, minutes, seconds, 0, 0);

                return t;
            }

            var input = ['1', time(1),
                         '11', time(11),
                         '111', time(1, 11),
                         '1234', time(12, 34),
                         '12345', time(1, 23, 45),
                         '123456', time(12, 34, 56),
                         '441234', time(4, 41, 23),
                         '4412345', time(4, 41, 23),
                         '44123456', time(4, 41, 23),
                         '441234567', time(4, 41, 23),
                         '446161', time(4, 46, 16),
                         '6666', time(7, 07, 0),
                         '66666', time(7, 07, 06),
                         '666666', time(7, 07, 06),
                         '6666666', time(7, 07, 06),
                         '46', time(5),
                         ':1', time(10),
                         ':2', time(20),
                         ':3', time(3),
                         ':4', time(4),
                         ':5', time(5),
                         ':6', time(6),
                         ':7', time(7),
                         ':8', time(8),
                         ':9', time(9),
                         ':12', time(12),
                         ':123', time(1, 23),
                         ':1234', time(12, 34),
                         ':12345', time(1, 23, 45),
                         ':123456', time(12, 34, 56),
                         ':1234567', time(12, 34, 56),
                         ':61', time(6, 10),
                         ':62', time(6, 20),
                         ':1261', time(13, 1),
                         ':1271', time(13, 11),
                         '1:7', time(1, 7),
                         '2:8', time(2, 8),
                         '3:9', time(3, 9),
                         '1:1', time(1, 10),
                         '6:1', time(6, 10),
                         '1:6', time(2),
                         '7:1', time(7, 10),
                         '8:60', time(9),
                         '8:59', time(8, 59),
                         '7:35', time(7, 35),
                         '3:45', time(3, 45),
                         '10:7', time(10, 7),
                         '21:08', time(21, 8),
                         '10:10', time(10, 10),
                         '10:60', time(10, 60),
                         '10:1', time(10, 10),
                         '10:3', time(10, 30),
                         '10:5', time(10, 50),
                         '12:77', time(13, 17),
                         '6:0660', time(6, 7),
                         '6:032', time(6,3,20),
                         '1:23', time(1, 23),
                         '2:345', time(2, 34, 50),
                         '3:4567', time(3, 46, 7),
                         '4:56012', time(4, 56, 1),
                         '123:4', time(1, 23, 4),
                         '1234:5', time(12, 34, 5),
                         '123:45', time(1, 23, 45),
                         '1234:56', time(12, 34, 56),
                         '1:2:3', time(1,2,3),
                         '1:2:30', time(1,2,30),
                         '10:2:30', time(10,2,30),
                         '1:20:30', time(1,20,30),
                         '11:15:03', time(11,15,3),
                         undefined, null];

            for (k = 0, n = input.length; k < n; k = k+2) {
                value = input[k];
                expected = input[k+1];
                t = instance.parse(value);
                result = t ? time(t.getHours(), t.getMinutes(), t.getSeconds()) : false;
                parsed = result ? result.toLocaleTimeString() : false;
                message = expected ? expected.toLocaleTimeString() : 'false';

                equal(result ? result.toLocaleTimeString() : null, expected ? expected.toLocaleTimeString() : null, 'Input: ' + value);
                // ok(result >= expected && result <= expected, 'Input:' + value + ' | Parsed: ' + parsed + ' | Expected: ' + message);
            }
        });

        test('format', function() {
            var timepicker = $('#timepicker').timepicker(),
                instance = timepicker.timepicker(),
                time = new Date(1988, 8, 24, 19, 30, 0),
                k, n, formats, result, format, expected;

            formats = [['hh:mm:ss p', '07:30:00 PM'],
                       ['hh:mm:ss a', '07:30:00 pm'],
                       ['HH:mm:ss', '19:30:00'],
                       ['h:m:s p', '7:30:0 PM'],
                       ['h:m:s a', '7:30:0 pm'],
                       ['H:m:s', '19:30:0']];

            for (k = 0, n = formats.length; k < n; k = k + 1) {
                format = formats[k][0];
                expected = formats[k][1];
                result = instance.format(time, format);

                ok(result === expected, 'Object: ' + time.toLocaleTimeString() + ' | Format: ' + format + ' | Result: ' + result);
            }
        });

        test('getTime/setTime', function() {
            var element = $('#timepicker').timepicker({
                    'timeFormat': 'hh:mm p'
                }),
                instance = element.timepicker(),
                date = new Date(0,0,0,12,50,34),
                object = null;

            object = instance.setTime(date);
            ok(object.jquery, 'setTime() method returns a jQuery instance');
            equal(element.val(), '12:50 PM', 'passing a Date object to setTime.');

            equal(instance.getTime().toLocaleTimeString(), date.toLocaleTimeString(), 'getTime return the time set by setTime using a Date object.');

            element.timepicker('setTime', '1:20p');
            equal(element.val(), '01:20 PM', 'passing a string to setTime.');

            date = new Date(0,0,0,13,20,0);
            equal(instance.getTime().toLocaleTimeString(), date.toLocaleTimeString(), 'getTime returns the time set by setTime using a string.');

            date = new Date(0,0,0,13,30,0);
            equal(element.val('1:30 PM').timepicker('getTime').toLocaleTimeString(), date.toLocaleTimeString(), 'getTime returns the time set by jQuery.fn.val.');
        });

        test('option', function() {
            var element = $('#timepicker').timepicker({
                    'timeFormat': 'hh:mm p'
                }),
                instance = element.timepicker();

            instance.setTime('11:40');

            equal(instance.option('timeFormat'), 'hh:mm p', 'timeFormat: value succesfully retrieved (instance).');

            instance.option('timeFormat', 'h p');
            equal(instance.format(instance.getTime()), '11 AM', 'timeFormat: value succesfully updated (instance).');
            equal(element.val(), '11 AM', 'timeFormat: input field value was properly updated with the new format (instance).');

            equal(element.timepicker('option', 'timeFormat'), 'h p', 'timeFormat: value succesfully retrieved (plugin).');

            element.timepicker('option', 'timeFormat', 'h:mm p');
            equal(instance.format(instance.getTime()), '11:40 AM', 'timeFormat: value succesfully updated (plugin).');
            equal(element.val(), '11:40 AM', 'timeFormat: input field value was properly updated with the new format (plugin).');
        });


        module('TimePicker Options', { teardown: teardown });

        test('defaultTime:string', function() {
            var timepicker = $('#timepicker').timepicker({ defaultTime: '10p', timeFormat: 'hh:mm:ss a' });

            equal(timepicker.val(), '10:00:00 pm', 'A default time can be passed in the `defaultTime` option.');
            deepEqual(timepicker.timepicker('getTime'), new Date(0, 0, 0, 22, 0, 0), 'A default time can be passed in the `defaultTime` option.');
        });

        test('defaultTime:Date', function() {
            var now = new Date(), timepicker;

            now = new Date(0, 0, 0, now.getHours(), now.getMinutes(), now.getSeconds());
            timepicker = $('#timepicker').timepicker({ defaultTime: now, timeFormat: 'hh:mm:ss a' });

            deepEqual(timepicker.timepicker('getTime'), now, 'A default time can be passed in the `defaultTime` option.');
        });

        test('defaultTime:now', function() {
            var now = new Date(), timepicker;

            now = new Date(0, 0, 0, now.getHours(), now.getMinutes(), now.getSeconds());
            timepicker = $('#timepicker').timepicker({ defaultTime: 'now', timeFormat: 'hh:mm:ss a' });

            deepEqual(timepicker.timepicker('getTime'), now, 'A default time can be passed in the `defaultTime` option.');
        });

        test('minTime', function() {
            var timepicker, instance;

            timepicker = $('#timepicker').timepicker({ minTime: '2p' });
            instance = timepicker.timepicker();

            timepicker.val('13:59:59');
            equal(instance.getTime(), null, 'An invalid time cannot be set using $.fn.val().');

            timepicker.val('2p');
            deepEqual(instance.getTime(), new Date(0, 0, 0, 14, 0, 0), 'A time equal to minTime can be set using $.fn.val().');
        });

        test('interval', function() {
            var timepicker, instance;
            timepicker = $('#timepicker').timepicker({ interval:30, timeFormat: 'h' });
            instance = timepicker.timepicker();

            ok(instance.options.interval === 30, 'interval was properly set to 30.');
            timepicker.focus();
            ok(instance.options.interval === 60, 'interval was changed to 60, because the timeFormat does not include m or mm.');
            timepicker.blur();

            instance.option('interval', 90);
            timepicker.focus();
            ok(instance.options.interval === 120, 'interval was changed to 120, after the interval was updated, because the timeFormat does not include m or mm.');
            timepicker.blur();

            instance.option('interval', 45);
            instance.option('timeFormat', 'hh:mm');
            timepicker.focus();
            ok(instance.options.interval === 45, 'interval was not changed to 60, after the interval and timeFormat were updated, because the timeFormat includes m or mm.');
            timepicker.blur();
        });

        test('dropdown', function() {
            var timepicker, instance;
            timepicker = $('#timepicker').timepicker({ dropdown: false });
            instance = timepicker.timepicker();
            timepicker.focus();
            ok(instance.closed(), 'dropdown is not opened if dropdown was set to false.');
            timepicker.blur();

            instance.option('dropdown', true);
            timepicker.focus();
            ok(!instance.closed(), 'dropdown is opened if dropdown was set to true.');
        });

        test('callbacks', function() {
            var expected = 1, timepicker;

            timepicker = $('#timepicker').timepicker({
                change: function(/*time*/) {
                    ok(true, 'change() callback executed.');
                    start();
                }
            });

            expect(expected);
            stop();

            timepicker.val('46').change();
        });



        module('TimePicker Event Handlers');

        test('opened/closed', function() {
            var timepicker = $('#timepicker').timepicker(),
                instance = timepicker.timepicker(),
                selected = null,
                expected = 8;

            expect(expected);
            stop();

            system.queue('test', []);
            system.queue('test', function(next) {
                ok(instance.closed(), 'TimePicker starts closed.');
                timepicker.focus();
                next();
            });

            system.delay(delay, 'test');
            system.queue('test', function(next) {
                ok(!instance.closed(), 'TimePicker opens when input field gains focus.');
                timepicker.simulate('keydown', {keyCode: 65});
                next();
            });

            system.delay(delay, 'test');
            system.queue('test', function(next) {
                ok(instance.closed(), 'TimePicker is closed after the \'a\' key is pressed.');
                timepicker.simulate('keydown', {keyCode: $.TimePicker.prototype.keyCode.DOWN});
                next();
            });

            system.delay(delay, 'test');
            system.queue('test', function(next) {
                ok(!instance.closed(), 'TimePicker opens when DOWN key is pressed.');
                selected = instance.selected();
                timepicker.simulate('keydown', {keyCode: 65});
                timepicker.simulate('keydown', {keyCode: $.TimePicker.prototype.keyCode.UP});
                next();
            });

            system.delay(delay, 'test');
            system.queue('test', function(next) {
                ok(!instance.closed(), 'TimePicker opens when UP key is pressed.');
                selected = instance.selected();
                timepicker.simulate('keydown', {keyCode: $.TimePicker.prototype.keyCode.ENTER});
                next();
            });

            system.delay(delay, 'test');
            system.queue('test', function(next) {
                ok(instance.closed(), 'TimePicker is closed after an item has been selected pressing ENTER key.');
                ok(selected !== null, 'An element is selected after the DOWN key is pressed.');
                ok(selected ? timepicker.val() === selected.text() : false, 'The value in the input field is the text of the selected item.');
                next();
            }, 50);

            system.queue('test', function(/*next*/) { start(); }).dequeue('test');
        });
    });
})(jQuery);


test/playground.js
/*global LazyLoad: false, jQuery: false*/
/*jslint indent: 4*/

(function() {
    var jquery, ui, timepicker, test, js, css,
        parts = location.search.replace(/^\?/, '').split('&'),
        n = parts.length, params = {};

    for (var i = 0; n > 1 && i < n; i = i + 1) {
        parts[i] = parts[i].split('=');
        params[parts[i][0]] = parts[i][1];
    }

    jquery = params['jquery-version'] || '1.7.1';
    ui = params['jquery-ui-version'] || '1.8.2';
    timepicker = params.timepicker || 'normal';
    test = params.test || 'core';

    js = ['../external/jquery-'+jquery+'.min.js',
          '../external/jquery-ui-'+ui+'/jquery-ui-'+ui+'.custom.min.js',
          '../external/jquery.simulate.js',
          '../external/qunit.js',
          '../jquery.timepicker'+(timepicker === 'normal' ? '' : '.min')+'.js',
          'timepicker-'+test+'-test.js'];

    css = ['../external/jquery-ui'+ui+'/jquery-ui-'+ui+'.custom.css',
           '../jquery.timepicker.css'];

    LazyLoad.css(css, function() {
        // console.log('CSS Ready!');
    });

    LazyLoad.js(js, function() {
        // console.log('JS Ready!');
        if (jQuery.fn.timepicker.test) {
            jQuery.fn.timepicker.test();
        }
    });
})();

tests/parseTime.js
(function($) {

module('$.timepicker.parseTime');

var now = new Date(), k;
function time(hours, minutes, seconds) {
    hours = hours || 0;minutes = minutes || 0;seconds = seconds || 0;
    var time = new Date();
    time.setTime(now.valueOf());
    time.setHours(hours, minutes, seconds, 0, 0);
    return time;
}

function parseTimeTest(value, expected) {
    test('parseTime(\'' + value + '\')', 1, function() {
        var t = $.fn.timepicker.parseTime(value),
            result = null, parsed = '', expectedMessage;

        result = t ? time(t.getHours(), t.getMinutes(), t.getSeconds()) : false;
        parsed = result ? result.toTimeString() : false;
        expectedMessage = expected ? expected.toTimeString() : 'false';

        ok(result >= expected && result <= expected, 'Parsed: ' + parsed + ' | Expected: ' + expectedMessage);
    });
}

 var input = ['1', time(1),
              '11', time(11),
              '111', time(1, 11),
              '1234', time(12, 34),
              '12345', time(1, 23, 45),
              '123456', time(12, 34, 56),
              '46', time(5),
              ':1', time(10),
              ':2', time(20),
              ':3', time(3),
              ':4', time(4),
              ':5', time(5),
              ':6', time(6),
              ':7', time(7),
              ':8', time(8),
              ':9', time(9),
              ':12', time(12),
              ':123', time(1, 23),
              ':1234', time(12, 34),
              ':12345', time(1, 23, 45),
              ':123456', time(12, 34, 56),
              ':1234567', time(12, 34, 56),
              ':61', time(6, 10),
              ':62', time(6, 20),
              ':1261', false,
              ':1271', false,
              '1:7', time(1, 7),
              '2:8', time(2, 8),
              '3:9', time(3, 9),
              '1:1', time(1, 10),
              '6:1', time(6, 10),
              '1:6', time(2),
              '7:1', time(7, 10),
              '8:60', time(9),
              '8:59', time(8, 59),
              '7:35', time(7, 35),
              '3:45', time(3, 45),
              '10:7', time(10, 7),
              '21:08', time(21, 8),
              '10:10', time(10, 10),
              '10:60', time(10, 60),
              '10:1', time(10, 10),
              '10:3', time(10, 30),
              '10:5', time(10, 50),
              '6:0660', time(6, 7),
              '6:032', time(6,3,20),
              '1:23', time(1, 23),
              '2:345', time(2, 34, 50),
              '3:4567', false,
              '4:56012', time(4, 56, 1),
              '123:4', time(1, 23, 4),
              '1234:5', time(12, 34, 5),
              '123:45', time(1, 23, 45),
              '1234:56', time(12, 34, 56),
              '1:2:3', time(1,2,3),
              '1:2:30', time(1,2,30),
              '10:2:30', time(10,2,30),
              '1:20:30', time(1,20,30),
              '11:15:03', time(11,15,3)];

for (k = 0; k < input.length; k = k+2) {
    parseTimeTest(input[k], input[k+1]);
}

})(jQuery);

resources/js/timepicker.page.js
/*global _gaq*/
$(function() {

    /* activate timepickers */

    $('#options-time-format').timepicker({ timeFormat: 'h:mm:ss p' });

    $('#options-time-constraints').timepicker({
        timeFormat: 'HH:mm:ss',
        minTime: '11:45:00',
        maxHour: 20,
        maxMinutes: 30,
        startTime: new Date(0,0,0,15,0,0),
        interval: 15
    });

    $('#options-change-event').timepicker({
        change: function(time) {
            // the input field
            var element = $(this), text;
            // get access to this TimePicker instance
            var timepicker = element.timepicker();
            text = 'Selected time is: ' + timepicker.format(time);
            element.siblings('span.help-line').text(text);
        }
    });

    /* track Google Analytics events */

    var trackOutboundLinks = function(href, category, action) {
        try {
            _gaq.push(['_trackEvent', category, action]);
        } catch (err) {
            // pass
        }

        setTimeout(function() {
            document.location.href = href;
        }, 100);
    };

    $('#download-button').click(function(e) {
        e.preventDefault();

        var href = $(this).attr('href');

        trackOutboundLinks(href, 'Outbound Links', href);
    });
});


resources/js/page.js
if ( typeof jQuery !== 'undefined' ) {
    (function($) {
        $('.intro .timepicker-without-dropdown').timepicker({
            dropdown: false
        });

        $('.intro .timepicker-with-dropdown').timepicker({
        });

        $(function () {
            var $section = $('.jt-configuration-demo');

            if ( $section.length === 0 ) {
                return;
            }

            var $form = $section.find('.jt-configuration-demo-form');

            $form.find('[data-toggle="tooltip"]').tooltip({
                html: true,
                container: $form,
            });

            var $timepicker = $section.find('.timepicker'),
                $textfields = $(),
                $checkboxes = $(),
                options = {
                    timeFormat: 'h:mm p',
                    interval: 60,
                    defaultTime: '11',
                    startTime: '10:00',
                    minTime: '10',
                    maxTime: '6:00pm',
                    dynamic: false,
                    dropdown: true,
                    scrollbar: true
                };

            $textfields = $textfields.add($form.find('[name="timeFormat"]').val(options.timeFormat));
            $textfields = $textfields.add($form.find('[name="interval"]').val(options.interval));
            $textfields = $textfields.add($form.find('[name="defaultTime"]').val(options.defaultTime));
            $textfields = $textfields.add($form.find('[name="startTime"]').val(options.startTime));
            $textfields = $textfields.add($form.find('[name="minTime"]').val(options.minTime));
            $textfields = $textfields.add($form.find('[name="maxTime"]').val(options.maxTime));

            $textfields.each(function() {
                var $field = $(this);
                $field.val(options[ $field.attr('name') ]);
            }).change(function() {
                var $field = $(this);

                options[ $field.attr('name') ] = $field.val();

                updateTimepickerOptions();
            });

            $checkboxes = $checkboxes.add($form.find('[name="dynamic"]'));
            $checkboxes = $checkboxes.add($form.find('[name="dropdown"]'));
            $checkboxes = $checkboxes.add($form.find('[name="scrollbar"]'));

            $checkboxes.each(function() {
                var $field = $(this);
                $field.prop('checked', options[ $field.attr('name') ]);
            }).change(function() {
                var $field = $(this);

                options[ $field.attr('name') ] = $field.prop('checked');

                updateTimepickerOptions();
            });

            updateTimepickerOptions();

            function updateTimepickerOptions() {
                $timepicker.timepicker('destroy').timepicker(options);
                $section.find('.jt-configuration-demo-example-code-container code').html(formatOptions(options));
            }

            function formatOptions(options) {
                var code = '<span class="nx">$</span><span class="p">(</span><span class="s1">&#39;.timepicker&#39;</span><span class="p">).</span><span class="nx">timepicker</span><span class="p">({</span>' + "\n";

                if (options.timeFormat) {
                    code += formatStringOption('timeFormat', options.timeFormat) + ",\n";
                }

                if (options.interval) {
                    code += formatIntegerOption('interval', options.interval) + ",\n";
                }

                if (options.minTime) {
                    code += formatStringOption('minTime', options.minTime) + ",\n";
                }

                if (options.maxTime) {
                    code += formatStringOption('maxTime', options.maxTime) + ",\n";
                }

                if (options.defaultTime) {
                    code += formatStringOption('defaultTime', options.defaultTime) + ",\n";
                }

                if (options.startTime) {
                    code += formatStringOption('startTime', options.startTime) + ",\n";
                }

                code += formatBooleanOption('dynamic', options.dynamic) + ",\n";
                code += formatBooleanOption('dropdown', options.dropdown) + ",\n";
                code += formatBooleanOption('scrollbar', options.scrollbar) + "\n";

                return code + '<span class="p">});</span></code>';
            }

            function formatStringOption(name, value) {
                return "&nbsp;&nbsp;&nbsp;&nbsp;" + '<span class="nx">' + name + '</span><span class="o">:</span> <span class="s1">&#39;' + value + '&#39;</span><span class="p"></span>';
            }

            function formatIntegerOption(name, value) {
                return "&nbsp;&nbsp;&nbsp;&nbsp;" + '<span class="nx">' + name + '</span><span class="o">:</span> <span class="mi">' + value + '</span><span class="p"></span>';
            }

            function formatBooleanOption(name, value) {
                return "&nbsp;&nbsp;&nbsp;&nbsp;" + '<span class="nx">' + name + '</span><span class="o">:</span> <span class="kc">' + (value ? 'true' : 'false') + '</span>';
            }
        });
    })(jQuery);
}
